src/java.base/share/classes/java/util/concurrent/CompletableFuture.java

```diff
@@ -2839,15 +2809,53 @@ public CompletableFuture<T> orTimeout(long timeout, TimeUnit unit) {
      */
     public CompletableFuture<T> completeOnTimeout(T value, long timeout,
                                                   TimeUnit unit) {
-        if (unit == null)
-            throw new NullPointerException();
-        if (result == null)
-            whenComplete(new Canceller(Delayer.delay(
-                                           new DelayedCompleter<T>(this, value),
-                                           timeout, unit)));
+        arrangeTimeout(unit.toNanos(timeout),
+                       new Timeout<T>(this, value, false));
         return this;
     }
 
+    /** Action to complete (possibly exceptionally) on timeout */
+    static final class Timeout<U> implements Runnable {
+        final CompletableFuture<U> f;
+        final U value;
+        final boolean exceptional;
+        Timeout(CompletableFuture<U> f, U value, boolean exceptional) {
+            this.f = f; this.value = value; this.exceptional = exceptional;
+        }
+        public void run() {
+            if (f != null && !f.isDone()) {
+                if (exceptional)
+                    f.completeExceptionally(new TimeoutException());
+                else
+                    f.complete(value);
+            }
+        }
+    }
+
+    /** Action to cancel unneeded timeouts */
+    static final class Canceller implements BiConsumer<Object, Throwable> {
+        final Future<?> f;
+        Canceller(Future<?> f) { this.f = f; }
+        public void accept(Object ignore, Throwable ex) {
+            if (f != null)
```

Comment:
@DougLea Does it make sense to create a Canceller with a `null` `f`-parameter?

Reply:
The usual j.u.c answer: it is never null now, but if there were ever a case in which it were, it's a no-op.